yt/utilities/lib/particle_kdtree_tools.pyx

```diff
@@ -133,11 +133,18 @@ def generate_nn_list(np.float64_t[:] bounds, np.int64_t[:] dimensions,
     cdef int i, j, k, l, skip, xsize, ysize, zsize, npixels
     cdef BoundedPriorityQueue queue = BoundedPriorityQueue(n_neighbors, True)
     cdef np.float64_t x_min, x_max, y_min, y_max, z_min, z_max, dx, dy, dz
+    cdef np.int64_t[:,:,:,:] pixel_pids
+    cdef np.float64_t[:,:,:,:] pixel_distances
 
     # setting up the pixels to loop through
     xsize, ysize, zsize = dimensions[0], dimensions[1], dimensions[2]
     n_pixels = xsize*ysize*zsize
 
+    pixel_pids = np.zeros((xsize, ysize, zsize, n_neighbors),
+                          dtype=int)
+    pixel_distances = np.zeros((xsize, ysize, zsize, n_neighbors),
+                           dtype=float)
```

Comment:
These might be big arrays for a large dimensions. Do you think there's a way to either avoid allocating these or only allocate them if they're needed? For example, scatter smoothing doesn't need to allocate these arrays.

I haven't thought terribly carefully about this, but I think ine way to avoid allocating this would be to write a completely separate function for the gather smoothing that loops over voxels rather than particles. Then you only need to allocate an `n_neighbors` length 1D array that you can reuse for each pixel.

Reply:
That was pretty much what I had in mind. I was thinking about writing two seperate functions and having a wrapper that calls gather or scatter depending on the dataset properties. 

I'm not very knowledgeable with cython, but is there a better way to allocate memory of this kind of size? are these arrays contiguous in memory?